File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/tx/offline/OfflineSigningEnvelopeCodec.java
package org.hyperledger.iroha.android.tx.offline;

import org.hyperledger.iroha.android.norito.NoritoException;
import org.hyperledger.iroha.norito.NoritoCodec;

/**
 * Encodes and decodes {@link OfflineSigningEnvelope} instances using the shared Norito codec.
 */
public final class OfflineSigningEnvelopeCodec {

  private static final String DEFAULT_SCHEMA = "iroha.android.offline.Envelope.v1";
  private final String schemaName;
  private final OfflineSigningEnvelopeAdapter adapter = new OfflineSigningEnvelopeAdapter();

  public OfflineSigningEnvelopeCodec() {
    this(DEFAULT_SCHEMA);
  }

  public OfflineSigningEnvelopeCodec(final String schemaName) {
    this.schemaName = schemaName;
  }

  public byte[] encode(final OfflineSigningEnvelope envelope) throws NoritoException {
    try {
      return NoritoCodec.encode(envelope, schemaName, adapter);
    } catch (final Exception ex) {
      throw new NoritoException("Failed to encode offline signing envelope", ex);
    }
  }

  public OfflineSigningEnvelope decode(final byte[] encoded) throws NoritoException {
    try {
      return NoritoCodec.decode(encoded, adapter, schemaName);
    } catch (final Exception ex) {
      throw new NoritoException("Failed to decode offline signing envelope", ex);
    }
  }

  public String schemaName() {
    return schemaName;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/tx/offline/OfflineSigningEnvelopeCodec.kt
package org.hyperledger.iroha.sdk.tx.offline

import org.hyperledger.iroha.sdk.norito.NoritoCodec
import org.hyperledger.iroha.sdk.tx.norito.NoritoException

private const val DEFAULT_SCHEMA = "iroha.android.offline.Envelope.v1"

/** Encodes and decodes `OfflineSigningEnvelope` instances using the shared Norito codec. */
class OfflineSigningEnvelopeCodec @JvmOverloads constructor(
    @JvmField val schemaName: String = DEFAULT_SCHEMA,
) {
    private val adapter = OfflineSigningEnvelopeAdapter()

    @Throws(NoritoException::class)
    fun encode(envelope: OfflineSigningEnvelope): ByteArray {
        try {
            return NoritoCodec.encode(envelope, schemaName, adapter)
        } catch (ex: Exception) {
            throw NoritoException("Failed to encode offline signing envelope", ex)
        }
    }

    @Throws(NoritoException::class)
    fun decode(encoded: ByteArray): OfflineSigningEnvelope {
        try {
            return NoritoCodec.decode(encoded, adapter, schemaName)
        } catch (ex: Exception) {
            throw NoritoException("Failed to decode offline signing envelope", ex)
        }
    }
}
